optuna/storages/_rdb/storage.py

```diff
@@ -607,9 +607,16 @@ def _update_trial(
         """
 
         session = self.scoped_session()
+
+        # Lock all related tables with eager joins.
+        # TODO(hvy): Naive in terms of performance. Avoid joining with unncessary tables.
         trial_model = (
             session.query(models.TrialModel)
             .filter(models.TrialModel.trial_id == trial_id)
+            .options(orm.selectinload(models.TrialModel.params))
+            .options(orm.selectinload(models.TrialModel.values))
+            .options(orm.selectinload(models.TrialModel.user_attributes))
+            .options(orm.selectinload(models.TrialModel.system_attributes))
```

Comment:
I dumped the SQL queries related to this line. The queries are the same whether we add `selectinload` or not as follows: 

```sql
SELECT trials.trial_id AS trials_trial_id, trials.number AS trials_number, trials.study_id AS trials_study_id, trial
s.state AS trials_state, trials.value AS trials_value, trials.datetime_start AS trials_datetime_start, trials.datetime_complete AS trials_datetime_complete 
FROM trials 
WHERE trials.trial_id = %(trial_id_1)s FOR UPDATE
```

As can be seen in the example in the [reference](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html#selectin-eager-loading), the change is can be seen in the following SELECT queries.

```sql
# without selectinload
SELECT trial_user_attributes.trial_user_attribute_id AS trial_user_attributes_trial_user_attribute_id, trial_user_attributes.trial_id AS trial_user_attributes_trial_id, trial_user_attributes.`key` AS trial_user_attributes_key, trial_user_attributes.value_json AS trial_user_attributes_value_json 
FROM trial_user_attributes 
WHERE trial_user_attributes.trial_id = %(trial_id_1)s

# with selectionload
SELECT trial_user_attributes.trial_id AS trial_user_attributes_trial_id, trial_user_attributes.trial_user_attribute_id AS trial_user_attributes_trial_user_attribute_id, trial_user_attributes.`key` AS trial_user_attributes_key, trial_user_attributes.value_json AS trial_user_attributes_value_json 
FROM trial_user_attributes 
WHERE trial_user_attributes.trial_id IN (%(primary_keys_1)s) ORDER BY trial_user_attributes.trial_id```
```

So, I guess the `selectinload` option does not affect the record lock, but I may misunderstand the lock mechanism. What do you think of it?

Reply:
I used the following code to dump SQL queries: https://gist.github.com/toshihikoyanase/f839e9ab9b2cbba23366b635f12b79da

Reply:
Thanks, you're right. We probably need to do a proper "JOIN" (e.g. `orm.joinedload`) here.

Reply:
Updated. I verified that this properly locks backref tables. Sorry but could you check it again?

Note that we cannot do an `INNER JOIN` ([`innerjoin=True`](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html#sqlalchemy.orm.joinedload)) and must do an `LEFT OUTER JOIN` since backref tables don't necessarily have rows corresponding to the trial ID when we enter this section. Performance-wise, this change doesn't seem to have any noticeable impact, but I've only tested on a very trivial case.

Reply:
The change seems good to me, but I sometimes encountered the deadlock as follows:

```console
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')
[SQL: INSERT INTO trial_user_attributes (trial_id, `key`, value_json) VALUES (%(trial_id)s, %(key)s, %(value_json)s)]
[parameters: {'trial_id': 23, 'key': 'x', 'value_json': '7.048336196015672'}]
(Background on this error at: http://sqlalche.me/e/13/e3q8)

During handling of the above exception, another exception occurred:

optuna.exceptions.StorageInternalError: An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length. (The actual exception is as follows: OperationalError("(pymysql.err.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')"))
```

Do you think we need retry like #1509?

Reply:
Thanks for the additional information, this time the query seems to align with the expectation. It's still not entirely clear to me why the deadlocks were occur however. As for this issue, I changed the logic to not lock at all as they actually shouldn't be needed.

Reply:
Thank you for your update. I read the comment https://github.com/optuna/optuna/pull/1514#issuecomment-658001961, and I think your latest change sounds good to me.